test/broadcast.jl

```diff
@@ -300,7 +300,6 @@ import Base.Meta: isexpr
 # PR 16988
 @test Base.promote_op(+, Bool) === Int
 @test isa(broadcast(+, [true]), Array{Int,1})
-@test Base.promote_op(Float64, Bool) === Float64
```

Comment:
FWIW
```Julia
promote_op(op::Type, T) = Core.Inference.return_type(op, Tuple{T})
promote_op(op::Type, T1, T2) = Core.Inference.return_type(op, Tuple{T1,T2})
```
seems to be inferable and handles cases where the constructor specializes, e.g.
`promote_op(Array{Int64}, Int)` then correctly returns `Array{Int64,1}`, not just `Array{Int64}`.

Not that I'm too worried about improving the obsolete `promote_op`, but maybe you can employ this somehow to increase the chances of getting a leaftype in `_broadcast`?